### PR TITLE
Adds node ID caching layer to lookup service.

### PR DESCRIPTION
**Pull request by rolandpoulter**
https://groups.google.com/forum/#!topic/rackhd/CiRQhVdkE38

Attempting to improve logging performance by introducing caching to lookup-service. First cache miss results in a DB query, subsequent cache lookups for the same IP or MAC address will wait for the first query to respond unless the there is already a valid entry in the cache.

This should greatly reduce number DB queries by the LookupService.

### CODE REVIEW

File: spec/lib/services/lookup-spec.js [hunk 1]
```diff
@@ -34,7 +34,54 @@ describe('Lookup Service', function () {
...
+          lookupService.checkNodeIdCache('testAddress').then(spy1, spy1);
+          lookupService.checkNodeIdCache('testAddress').then(spy2, spy2);
+        });
+
+        it('should resolve pending cache checks once a value is assigned', function () {
+          lookupService.assignNodeIdCache('testAddress', 'nodeId');
+          setTimeout(function () {
+            expect(spy1.called).to.be.ok;
+            expect(spy2.called).to.be.ok;
+            done();
```
Comment (benbp) [inline spec/lib/services/lookup-spec.js]: Not passed in as an argument to the 'it' callback.

File: lib/services/lookup.js [hunk 2]
```diff
@@ -71,9 +71,64 @@ function lookupServiceFactory(
...
+    };
+
+    /**
+     * Check the cache for a Node ID by IP or MAC addres.
+     * @param  {String} cache key - IP or MAC address.
+     * @return {Promise|null} Returns a promise if the Node ID is cached,
+                              otherwise null.
+     */
+    LookupService.prototype.checkNodeIdCache = function (key) {
+      if (typeof this.nodeIdCache[key] === 'string') {
```
Comment (benbp) [inline lib/services/lookup.js]: Double spacing here.

File: spec/lib/services/lookup-spec.js (hunk 1 above)
Comment (rolandpoulter) [inline spec/lib/services/lookup-spec.js]: Already fixed this. Tests should pass now.

**Comment by rolandpoulter** [comment]
Uggh. -0.02% code coverage loss.

**Comment by benbp** [comment]
Two questions:

1. Is this missing cache refresh/expiration right now?
2. I assume that independent calls to a cache promise object from different places within the code will chain onto the shared reference to the object, yes? That way we avoid the promise multicast problem?

**Comment by DavidjohnBlodgett** [comment]
I think code coverage will always go down if you delete anything~ even if its replaced with updated code~

**Comment by rolandpoulter** [comment]
1) Correct, it doesn't try to populate the cache, nor does it every clear the cache. It should be trivial memory usage even with 1000s of nodes, but if it is a problem we can easily address it then.

2) I think the answer is yes. I'll double check with you.

**Comment by heckj** [comment]
@rolandpoulter any reason not to use a the `lrucache` library with some options for timeout and decay of the cache over time - seems like that would have a smaller overall memory footprint, where this could conceivably grow memory usage without bound under pathological testing conditions (which we've experienced before)

**Comment by rolandpoulter** [comment]
@heckj Nope, I didn't know about it. It should be pretty simple to change it to use `lru-cache` instead of a plain Object.

I'll update my PR soon.

**Comment by rolandpoulter** [comment]
@benbp @heckj ^^

**Comment by DavidjohnBlodgett** [comment]
:+1:

**Comment by heckj** [comment]
:+1:

**Comment by yyscamper** [comment]
@rolandpoulter :

The caching is good, I just wondering doesn't the mongodb have such caching capability? if so, in Mongodb or Waterline, we can have a common cache layer for all query.

And could you explain how you sync your cache with real DB data?

**Comment by heckj** [comment]
@yyscamper MongoDB (like any database system) includes some caching, this is intended to cache *before* we make a request to mongoDB so that we don't need to incur the cost of that call every time we're using the lookup service.

LRU is specifically set to expire out data, and this code is set to attempt the cache first, and load on a cache miss, with a limited set of values. there's no explicit sync otherwise, so this does introduce the possibility that the DB can be updated and the cache will return a previous value that's not strictly correct. In the lookup service, the strict atomicity isn't necessary and the eventual consistency is quite acceptable.

File: lib/services/lookup.js [hunk 3]
```diff
@@ -71,9 +73,76 @@ function lookupServiceFactory(
...
+     * Reset all cached Node IDs.
+     */
+    LookupService.prototype.resetNodeIdCache = function () {
+        if (this.nodeIdCache) {
+          this.nodeIdCache.reset();
+          return;
+        }
+
+        this.nodeIdCache = LRU({
+            max: 500
```
Comment (heckj) [inline lib/services/lookup.js]: @benbp should we also be setting a `maxAge` key here in using the cache with the lookup services to limit how stale a cache value can be?

File: lib/services/lookup.js [hunk 4]
```diff
@@ -71,9 +73,76 @@ function lookupServiceFactory(
...
+        this.resetNodeIdCache();
     }
 
     /**
+     * Reset all cached Node IDs.
+     */
+    LookupService.prototype.resetNodeIdCache = function () {
+        if (this.nodeIdCache) {
+          this.nodeIdCache.reset();
+          return;
```
Comment (anhou) [inline lib/services/lookup.js]: indentation is 2, fixed to 4

**Comment by rolandpoulter** [comment]
Fixed whitespace.

**Comment by rolandpoulter** [comment]
@heckj @benbp I think this is ready to be merged.

**Comment by heckj** [comment]
:+1: from me

**Comment by rolandpoulter** [comment]
@benbp `lru-cache` was not being used before, so upgrading the version should not cause anything to break. Just double checked that.

**Comment by benbp** [comment]
:+1:

**Pull request merged**